subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultPolymorphicDomainObjectContainerTest.groovy

```diff
@@ -355,25 +369,29 @@ class DefaultPolymorphicDomainObjectContainerTest extends AbstractPolymorphicDom
         def expectedSchema = [
             mike: "DefaultPolymorphicDomainObjectContainerTest.Person",
             fred: "DefaultPolymorphicDomainObjectContainerTest.Person",
-            alice: "DefaultPolymorphicDomainObjectContainerTest.Person", // TODO: Should be AgeAwarePerson
-            kate: "DefaultPolymorphicDomainObjectContainerTest.Person",
-            bob: "DefaultPolymorphicDomainObjectContainerTest.Person",
-            mary: "DefaultPolymorphicDomainObjectContainerTest.Person", // TODO should be AgeAwarePerson
-            john: "DefaultPolymorphicDomainObjectContainerTest.Person",
-            janis: "DefaultPolymorphicDomainObjectContainerTest.Person", // TODO could be AgeAwarePerson
-            robert: "DefaultPolymorphicDomainObjectContainerTest.Person" // TODO could be DefaultCtorNamedPerson
+            alice: "DefaultPolymorphicDomainObjectContainerTest.AgeAwarePerson",
+            edward: "DefaultPolymorphicDomainObjectContainerTest.Introvert",
+            kate: "DefaultPolymorphicDomainObjectContainerTest.DefaultPerson",
+            bob: "DefaultPolymorphicDomainObjectContainerTest.DefaultPerson",
+            mary: "DefaultPolymorphicDomainObjectContainerTest.DefaultAgeAwarePerson",
+            john: "DefaultPolymorphicDomainObjectContainerTest.DefaultPerson",
+            janis: "DefaultPolymorphicDomainObjectContainerTest.DefaultAgeAwarePerson",
+            robert: "DefaultPolymorphicDomainObjectContainerTest.DefaultCtorNamedPerson",
+            kevin: "DefaultPolymorphicDomainObjectContainerTest.Person",
```

Comment:
:question: this is now exposing implementation details that were not exposed before. If we ship with this and later restrict this it would be a breaking change. We need to discuss this further.

Reply:
One thing we could do for now is to revert back to the previous implementation, which just exposed the one static, public element type the container was set up for, BUT make an exception for the elements that implement the `HasPublicType` marker interface. I think it would be a good compromise for now.